Questions from https://github.com/DataTalksClub/data-engineering-zoomcamp/blob/main/cohorts/2023/week_1_docker_sql/homework.md

# 1 Knowing docker tags

Run the command to get information on Docker

`docker --help`

We are looking for the tag that has the tag *"Write the image ID to the file"*


In [9]:
!docker --help build | grep "Write the image ID to the file"

      --iidfile string          Write the image ID to the file


**Answer**: the tag we are looking for is `--dfile string`

# 2 Understanding docker first run

In [10]:
!docker run python:3.9 pip list

Package    Version
---------- -------
pip        22.0.4
setuptools 58.1.0
wheel      0.38.4
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.


# 3 Count Records

For the SQL exercises, I initially ingested the latest `.parquet` files from the NYC TLC Trip Record Data. However, I realized there are slightly more records in this dataset compared to the `.csv` files that were archived, so there was a possibility of numbers not matching up exactly. As a result I put in a separate `green_taxi_trip_csv` table containing values ingested from the csv.

In [15]:
from sqlalchemy import create_engine, text
engine = create_engine('postgresql://root:root@localhost:5432/ny_taxi')


In [24]:
count_records_query = text("""
SELECT count(*) 
FROM green_taxi_trips_csv
WHERE date(lpep_pickup_datetime)='2019-01-15' AND DATE(lpep_dropoff_datetime)='2019-01-15';
""")
response = engine.execute(count_records_query)

for record in response:
    print(record)

(20664,)


Well, this doesn't match up with any of the multiple choice answers. The closest one is 20689 which is off by 25. 

# 4 Largest Trip for each Day

In [26]:
count_records_query = text("""
SELECT 
    date(lpep_pickup_datetime),
    max(trip_distance) as max_trip_distance
FROM green_taxi_trips_csv
GROUP BY date(lpep_pickup_datetime)
ORDER BY max_trip_distance desc
LIMIT 3;
""")
response = engine.execute(count_records_query)

for record in response:
    print(record)

(datetime.date(2019, 1, 15), 117.99)
(datetime.date(2019, 1, 18), 80.96)
(datetime.date(2019, 1, 28), 64.27)


It looks like the longest trip was on `2019-01-15`.

# 5 The Number of Passengers

In 2019-01-01 how many trips had 2 and 3 passengers?

In [27]:
count_records_query = text("""
SELECT
    passenger_count,
    count(*)
FROM green_taxi_trips_csv
WHERE 
    date(lpep_pickup_datetime)='2019-01-01'
GROUP BY passenger_count
HAVING passenger_count in (2,3);
""")
response = engine.execute(count_records_query)

for record in response:
    print(record)

(2, 1282)
(3, 254)


On that day:

    - There were 1282 rides with 2 passengers.
    - There were 254 rides with 3 passengers.

# 6 Largest tip
For the passengers picked up in the Astoria Zone which was the drop off zone that had the largest tip? We want the name of the zone, not the id.

In [29]:
count_records_query = text("""
SELECT
    lpep_pickup_datetime,
    lpep_dropoff_datetime,
    total_amount,
    tip_amount,
    zpu."Zone" as "pickup_loc",
    zdo."Zone" as "dropoff_loc"
FROM green_taxi_trips_csv gtt 
    join zones zpu on gtt."PULocationID"=zpu."LocationID"
    join zones zdo on gtt."DOLocationID"=zdo."LocationID"
WHERE 
    zpu."Zone"='Astoria'
ORDER BY tip_amount desc
LIMIT 3;
""")
response = engine.execute(count_records_query)

for record in response:
    print(record)

(datetime.datetime(2019, 1, 26, 0, 50, 10), datetime.datetime(2019, 1, 26, 0, 50, 10), 94.8, 88.0, 'Astoria', 'Long Island City/Queens Plaza')
(datetime.datetime(2019, 1, 24, 13, 29, 59), datetime.datetime(2019, 1, 24, 13, 29, 59), 65.06, 30.0, 'Astoria', 'Central Park')
(datetime.datetime(2019, 1, 20, 23, 18, 5), datetime.datetime(2019, 1, 20, 23, 18, 5), 56.8, 25.0, 'Astoria', 'Jamaica')


The trip with the largest tip that departed from Astoria went to **Long Island City/Queens Plaza**